File: src/anchor_v1/cose.py
"""ANCHOR v1 — hand-rolled COSE_Sign1 (RFC 9052) over Ed25519.

The signed object is:

    COSE_Sign1 = [ body_protected : bstr, unprotected : map,
                   payload : bstr, signature : bstr ]

with Sig_structure = ["Signature1", body_protected, external_aad, payload].

Protocol hardening (fail-closed by design):

* Algorithm allowlist: only EdDSA (-8) is accepted. Any other ``alg`` value in
  the protected header is rejected before signature verification.
* The protected header must contain exactly {1: -8, 4: kid}. No extra labels.
* The unprotected header must be empty. Header parameters smuggled outside the
  integrity-protected header are a classic downgrade vector.
* CBOR is decoded with the strict deterministic decoder — non-canonical
  encodings of the COSE structure are rejected.
* The ``kid`` selects the verification key from an explicit caller-supplied
  trust set; unknown kids are rejected.

Zero new dependencies: Ed25519 comes from the ``cryptography`` library already
used by ``anchor_v1.crypto``.
"""

from __future__ import annotations

from typing import Mapping

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

from .cbor import CBORError, cbor_dumps, cbor_loads


class COSEError(ValueError):
    """Raised when a COSE_Sign1 object is malformed or fails verification."""


# COSE algorithm + header label registry values (RFC 9053 / RFC 9052).
ALG_EDDSA = -8
_LABEL_ALG = 1
_LABEL_KID = 4


def cose_sign_bytes(
    payload: bytes,
    sign_fn: "Callable[[bytes], bytes]",
    kid: bytes,
    external_aad: bytes = b"",
) -> bytes:
    """Create a COSE_Sign1 message, signing via a caller-supplied function.

    Same wire format as :func:`cose_sign`; the only difference is that the
    Ed25519 signature over the Sig_structure is produced by ``sign_fn``
    (e.g. ``Ed25519Signer.sign_bytes``) instead of a raw private-key object.
    """
    protected = cbor_dumps({_LABEL_ALG: ALG_EDDSA, _LABEL_KID: bytes(kid)})
    sig_structure = cbor_dumps(
        ["Signature1", protected, bytes(external_aad), bytes(payload)]
    )
    signature = sign_fn(sig_structure)
    if not isinstance(signature, (bytes, bytearray)) or len(signature) != 64:
        raise COSEError("sign_fn must return a 64-byte Ed25519 signature")
    return cbor_dumps([protected, {}, bytes(payload), signature])


def cose_sign(
    payload: bytes,
    private_key: Ed25519PrivateKey,
    kid: bytes,
    external_aad: bytes = b"",
) -> bytes:
    """Create a COSE_Sign1 message over ``payload``.

    ``kid`` is placed in the protected header as label 4 (bstr). The protected
    header is ``{1: -8, 4: kid}`` and nothing else.
    """
    if not isinstance(payload, (bytes, bytearray)):
        raise COSEError("payload must be bytes")
    if not isinstance(kid, (bytes, bytearray)) or len(kid) == 0:
        raise COSEError("kid must be a non-empty byte string")
    if not isinstance(external_aad, (bytes, bytearray)):
        raise COSEError("external_aad must be bytes")

    protected = cbor_dumps({_LABEL_ALG: ALG_EDDSA, _LABEL_KID: bytes(kid)})
    sig_structure = cbor_dumps(
        ["Signature1", protected, bytes(external_aad), bytes(payload)]
    )
    signature = private_key.sign(sig_structure)
    return cbor_dumps([protected, {}, bytes(payload), signature])


def _validate_cose_headers(protected: Any, unprotected: Any) -> tuple[int, bytes]:
    """Validate protected and unprotected COSE header maps (ATTACK-12 / ATTACK-15). Fail-closed."""
    if not isinstance(unprotected, dict):
        raise COSEError("unprotected header must be a map")
    if unprotected != {}:
        raise COSEError("unprotected header must be empty")
    if not isinstance(protected, dict):
        raise COSEError("protected header must be a map")
    if set(protected.keys()) != {_LABEL_ALG, _LABEL_KID}:
        raise COSEError("protected header must contain exactly alg and kid")
    alg = protected[_LABEL_ALG]
    if alg != ALG_EDDSA:
        raise COSEError(f"algorithm not allowed: {alg!r} (only EdDSA/-8)")
    kid = protected[_LABEL_KID]
    if not isinstance(kid, bytes):
        raise COSEError("kid must be a byte string")
    return alg, kid


def cose_verify(
    message: bytes,
    trusted_keys: Mapping[bytes, Ed25519PublicKey],
    external_aad: bytes = b"",
) -> tuple[bytes, bytes]:
    """Verify a COSE_Sign1 message. Returns ``(payload, kid)``.

    ``trusted_keys`` maps kid bytes to Ed25519 public keys. Raises COSEError on
    any malformed input, disallowed algorithm, unknown kid, or bad signature.
    """
    if not isinstance(message, (bytes, bytearray)):
        raise COSEError("COSE message must be bytes")
    try:
        outer = cbor_loads(bytes(message))
    except CBORError as exc:
        raise COSEError(f"malformed COSE_Sign1: {exc}") from exc

    if not isinstance(outer, list) or len(outer) != 4:
        raise COSEError("COSE_Sign1 must be an array of 4 elements")
    body_protected, unprotected, payload, signature = outer
    if not isinstance(body_protected, bytes):
        raise COSEError("body_protected must be a byte string")
    if not isinstance(unprotected, dict):
        raise COSEError("unprotected header must be a map")
    if not isinstance(payload, bytes):
        raise COSEError("payload must be a byte string")
    if not isinstance(signature, bytes):
        raise COSEError("signature must be a byte string")

    try:
        protected = cbor_loads(body_protected)
    except CBORError as exc:
        raise COSEError(f"malformed protected header: {exc}") from exc

    alg, kid = _validate_cose_headers(protected, unprotected)

    key = trusted_keys.get(kid)
    if key is None:
        raise COSEError("unknown kid: no trusted key")

    sig_structure = cbor_dumps(["Signature1", body_protected, bytes(external_aad), payload])
    try:
        key.verify(signature, sig_structure)
    except InvalidSignature as exc:
        raise COSEError("signature verification failed") from exc
    return payload, kid

File: rust/verifier/src/cose.rs
//! Hand-rolled COSE_Sign1 (RFC 9052) verification over Ed25519.
//!
//! Mirrors `src/anchor_v1/cose.py`. The signed object is:
//!
//! ```text
//! COSE_Sign1 = [ body_protected : bstr, unprotected : map,
//!                payload : bstr, signature : bstr ]
//! ```
//!
//! with `Sig_structure = ["Signature1", body_protected, external_aad, payload]`.
//!
//! Protocol hardening (fail-closed, matching the Python side):
//! * algorithm allowlist: only EdDSA (-8),
//! * protected header must contain exactly `{1: -8, 4: kid}` — no extra labels,
//! * unprotected header must be empty,
//! * CBOR is decoded with the strict deterministic decoder.

use crate::cbor::{self, Integer, Value};
use crate::error::{Error, Result};
use ed25519_dalek::{Signature, VerifyingKey};

/// COSE algorithm + header label registry values (RFC 9053 / RFC 9052).
pub const ALG_EDDSA: i128 = -8;
const LABEL_ALG: i128 = 1;
const LABEL_KID: i128 = 4;

/// Verify a COSE_Sign1 message. Returns `(payload, kid)`.
///
/// `trusted_keys` maps kid bytes to 32-byte Ed25519 public keys. Fails closed
/// on any malformed input, disallowed algorithm, unknown kid, or bad signature.
pub fn verify(
    message: &[u8],
    trusted_keys: &[(Vec<u8>, [u8; 32])],
    external_aad: &[u8],
) -> Result<(Vec<u8>, Vec<u8>)> {
    let outer = cbor::loads(message)
        .map_err(|e| Error::new("malformed-cose", format!("malformed COSE_Sign1: {e}")))?;

    let items = outer.as_array().ok_or_else(|| {
        Error::new("malformed-cose", "COSE_Sign1 must be an array of 4 elements")
    })?;
    if items.len() != 4 {
        return Err(Error::new(
            "malformed-cose",
            "COSE_Sign1 must be an array of 4 elements",
        ));
    }
    let body_protected = items[0].as_bytes().ok_or_else(|| {
        Error::new("malformed-cose", "body_protected must be a byte string")
    })?;
    let unprotected = &items[1];
    let payload = items[2].as_bytes().ok_or_else(|| {
        Error::new("malformed-cose", "payload must be a byte string")
    })?;
    let signature = items[3].as_bytes().ok_or_else(|| {
        Error::new("malformed-cose", "signature must be a byte string")
    })?;
    if signature.len() != 64 {
        return Err(Error::new(
            "bad-signature-length",
            "signature must be a 64-byte Ed25519 signature",
        ));
    }

    let protected = cbor::loads(body_protected)
        .map_err(|e| Error::new("malformed-protected", format!("malformed protected header: {e}")))?;
    let (_alg, kid) = validate_headers(&protected, unprotected)?;

    let key_bytes = trusted_keys
        .iter()
        .find(|(k, _)| k.as_slice() == kid)
        .map(|(_, pk)| pk)
        .ok_or_else(|| Error::new("unknown-kid", "unknown kid: no trusted key"))?;
    let key = VerifyingKey::from_bytes(key_bytes)
        .map_err(|_| Error::new("bad-key", "trusted key is not a valid Ed25519 public key"))?;

    let sig_structure = cbor::dumps(&Value::Array(vec![
        Value::Text("Signature1".to_owned()),
        Value::Bytes(body_protected.to_vec()),
        Value::Bytes(external_aad.to_vec()),
        Value::Bytes(payload.to_vec()),
    ]))?;
    let sig = Signature::from_slice(signature)
        .map_err(|_| Error::new("bad-signature", "malformed Ed25519 signature"))?;
    key.verify_strict(&sig_structure, &sig)
        .map_err(|_| Error::new("bad-signature", "signature verification failed"))?;

    Ok((payload.to_vec(), kid.to_vec()))
}

/// Validate protected/unprotected headers (ATTACK-12 / ATTACK-15). Fail-closed.
fn validate_headers(protected: &Value, unprotected: &Value) -> Result<(i128, Vec<u8>)> {
    let up = unprotected.as_map().ok_or_else(|| {
        Error::new("bad-headers", "unprotected header must be a map")
    })?;
    if !up.is_empty() {
        return Err(Error::new("bad-headers", "unprotected header must be empty"));
    }
    let p = protected.as_map().ok_or_else(|| {
        Error::new("bad-headers", "protected header must be a map")
    })?;
    if p.len() != 2 {
        return Err(Error::new(
            "bad-headers",
            "protected header must contain exactly alg and kid",
        ));
    }
    let mut alg: Option<i128> = None;
    let mut kid: Option<Vec<u8>> = None;
    for (k, v) in p {
        match k {
            Value::Int(Integer::Small(n)) if *n == LABEL_ALG => alg = v.as_int(),
            Value::Int(Integer::Small(n)) if *n == LABEL_KID => {
                kid = v.as_bytes().map(|b| b.to_vec())
            }
            _ => {
                return Err(Error::new(
                    "bad-headers",
                    "protected header must contain exactly alg and kid",
                ))
            }
        }
    }
    let alg = alg.ok_or_else(|| {
        Error::new("bad-headers", "protected header must contain exactly alg and kid")
    })?;
    let kid = kid.ok_or_else(|| {
        Error::new("bad-headers", "protected header must contain exactly alg and kid")
    })?;
    if alg != ALG_EDDSA {
        return Err(Error::new(
            "bad-algorithm",
            format!("algorithm not allowed: {alg} (only EdDSA/-8)"),
        ));
    }
    Ok((alg, kid))
}
